examples/example_vm/example_vm.cpp

```diff
@@ -0,0 +1,300 @@
+
+    /// Pops an item from the top of the stack.
+    evmc_uint256be pop() { return *--pointer; }
+
+    /// Pushes an item to the top of the stack.
+    void push(evmc_uint256be value) { *pointer++ = value; }
+};
+
+/// The Example VM memory representation.
+struct Memory
+{
+    size_t size = 0;          ///< The current size of the memory.
+    uint8_t data[1024] = {};  ///< The fixed-size memory buffer.
+
+    /// Stores the given value bytes in the memory at given index.
+    /// The Memory::size is updated accordingly.
+    /// The memory buffer bounds are not checked.
+    void set(size_t index, const uint8_t* value_data, size_t value_size)
+    {
+        std::memcpy(&data[index], value_data, value_size);
+        size_t new_size = index + value_size;
+        if (new_size > size)
+            size = new_size;
+    }
+};
+
+/// The example implementation of the evmc_vm::execute() method.
+static evmc_result execute(evmc_vm* instance,
+                           const evmc_host_interface* host,
+                           evmc_host_context* context,
+                           enum evmc_revision rev,
+                           const evmc_message* msg,
+                           const uint8_t* code,
+                           size_t code_size)
+{
+    ExampleVM* vm = static_cast<ExampleVM*>(instance);
+
+    if (vm->verbose > 0)
+        std::puts("execution started\n");
+
+    int64_t gas_left = msg->gas;
+    Stack stack;
+    Memory memory;
+
+    for (size_t pc = 0; pc < code_size; pc++)
+    {
+        // Check remaining gas, assume each instruction costs 1.
+        gas_left -= 1;
+        if (gas_left < 0)
+            return evmc_make_result(EVMC_OUT_OF_GAS, 0, nullptr, 0);
+
+        switch (code[pc])
+        {
+        default:
+            return evmc_make_result(EVMC_UNDEFINED_INSTRUCTION, 0, nullptr, 0);
+
+        case OP_STOP:
+            return evmc_make_result(EVMC_SUCCESS, gas_left, nullptr, 0);
+
+        case OP_ADD:
+        {
+            uint8_t a = stack.pop().bytes[31];
+            uint8_t b = stack.pop().bytes[31];
+            uint8_t sum = static_cast<uint8_t>(a + b);
+            evmc_uint256be value = {};
+            value.bytes[31] = sum;
+            stack.push(value);
+            break;
+        }
+
+        case OP_ADDRESS:
+        {
+            evmc_address address = msg->destination;
+            evmc_uint256be value = {};
+            std::memcpy(&value.bytes[12], address.bytes, sizeof(address.bytes));
+            stack.push(value);
+            break;
+        }
+
+        case OP_CALLDATALOAD:
+        {
+            size_t offset = stack.pop().bytes[31];
+            evmc_uint256be value = {};
+
+            size_t copy_size = 0;
+            if (offset < msg->input_size)
+                copy_size = msg->input_size - offset;  // Partial copy.
+            if (copy_size > sizeof(value))
+                copy_size = sizeof(value);  // Full copy.
```

Comment:
A bit confusing part, maybe
```suggestion
            if (offset < msg->input_size)
              copy_size = std::min(sizeof(value), msg->input_size - offset)
```
or
```suggestion
  if (offset >= msg->input_size)
    stack.push(0);
  else
  {
      evmc_uint256be value = {};
      size_t copy_size = std::min(sizeof(value), msg->input_size - offset)
      ...
  }
```

Reply:
Changed